### PR TITLE
Test case and proposed fix for #558

### LINKED ISSUE #558: Expected object type [...] to exist for union
## Description

Suppose we have a union U = Foo | Bar and GraphQL type Bar is mapped to the Java interface Bar. In this case the concrete value for type Bar must be a class that implements Bar, i.e.:
```java
public interface Bar { ... }
public class BarImpl implements Bar { ... }

// and then a resolver method somewhere (GraphQL signature is `fooOrBar: U!`):
public Object getFooOrBar() { return new BarImpl(...); }
```
The UnionTypeResolver supports this scenario due to [PR 140](https://github.com/graphql-java-kickstart/graphql-java-tools/pull/140), which looks at the direct interfaces of the value.

However, if the type hierarchy is a bit more complex it will fail. Suppose we have an additional intermediate interface:
```java
public interface Bar { ... }
public interface BarEntity { ... }
public class BarEntityImpl implements BarEntity { ... }

// and then a resolver method somewhere (GraphQL signature is `fooOrBar: U!`)
public Object getFooOrBar() { return new BarEntityImpl(...); }
```
In this case the resolver will fail to resolve the type, since GraphQL type Bar is mapped to interface Bar, and that's not a direct superinterface of BarEntityImpl.

### Expected behavior

I would expect class BarEntityImpl to succesfully resolve to GraphQL type Bar, since it implements (indirectly) interface Bar.

### Actual behavior

Resolving fails with error "Expected object type with name 'BarEntityImpl' to exist for union 'U', but it doesn't!".

### Steps to reproduce the bug

Pull request #559 reproduces this issue.

### LINKED ISSUE #588: Is it possible to return a ResponseEntity<T> from a resolver
I'm using graphql-java with spring boot.
My schema looks like this.
type Query{
    userById(id: String): User
}
type User {
    id: String
    name: String
    email: String
}
I have a QueryResolver class that implements GraphQLQueryResolver. My get method inside this class looks like this.
public User getUserById(String id) {
   return someClient.getUserById(id).getBody(); //getUserById inside someClient returns a ResponseEntity<User>
}
What I need inside my QueryResolver is to return a ResponseEntity<User> instead of a User. So it'll look like the code below
public ResponseEntity<User>  getUserById(String id) {
   return someClient.getUserById(id); 
}
The above code returns a schema error saying org.springframework.http.ResponseEntity<User> cannot be mapped to a GraphQL type! Since GraphQL-Java deals with erased types at runtime, only non-parameterized classes can represent a GraphQL type. What does this mean? Is it possible to return a ResponseEntity<> from a resolver?

### PR DESCRIPTION
**Pull request by marcus-nl**
Fixes #558

## Checklist

- [x] Pull requests follows the [contribution guide](https://github.com/graphql-java-kickstart/graphql-java-tools/wiki/Contribution-guide)
- [x] New or modified functionality is covered by tests

## Description
DictionaryTypeResolver now considers the full type hierarchy -- superinterfaces in addition to superclasses.

### CODE REVIEW

File: src/test/kotlin/graphql/kickstart/tools/EndToEndSpecHelper.kt [hunk 1]
```diff
@@ -410,6 +434,11 @@ class MockPart(private val name: String, private val content: String) : Part {
     override fun delete() = throw IllegalArgumentException("Not supported")
 }
 
+interface Bar { val name: String }
+interface BarEntity : Bar { val id: String}
+class BarEntityImpl(override val id: String, override val name: String) : BarEntity
+class OutOfBeerError(val msg: String)
```
Comment (oryan-block) [inline src/test/kotlin/graphql/kickstart/tools/EndToEndSpecHelper.kt:440]: Can we think of a better example hierarchy here? Maybe `animal -> pet -> dog`?

**Review by oryan-block — CHANGES_REQUESTED** [review]
Thanks for the fix. I have one request.

File: src/test/kotlin/graphql/kickstart/tools/EndToEndSpecHelper.kt [hunk 2]
```diff
@@ -314,6 +328,16 @@ class Query : GraphQLQueryResolver, ListListResolver<String>() {
     fun throwsIllegalArgumentException(): String {
         throw IllegalArgumentException("Expected")
     }
+
+    fun allBars(): List<Bar> {
+        return listOf(BarEntityImpl("123", "Bar Name"))
```
Comment (oryan-block) [inline src/test/kotlin/graphql/kickstart/tools/EndToEndSpecHelper.kt:333]: This method can be converted to an expression body.

**Comment by marcus-nl** [comment]
Updated. 
(I could not use `animal -> pet -> dog`, since Animal was already defined in the same package in TestInterfaces.kt)

**Review by oryan-block — APPROVED** [review]

**Pull request merged**